File: original/simulation.m
% ====== simulation.m (Appendix A reconstruction) ======
%
% This script reproduces the simulation described in Chapter 5 and
% Appendix A of "Impact of an Infinite Cylinder".

close all
clear
clc

LxCGS = 160;      % Domain size in cm
lCGS = 1;         % Length of the pressure field in cm
m = 2^10;         % Number of spatial intervals
TauCGS = 0.5;     % Duration of surface forcing
TfCGS = TauCGS;   % Length of simulation
Nt = 150;         % Number of temporal intervals
rhoCGS = 1;       % Density in CGS
sigmaCGS = 70;    % Surface tension in CGS
gCGS = 980;       % Gravity in CGS

% ====== UNITS ======

L = lCGS;                 % Units of length
T = TauCGS;               % Units of time
M = rhoCGS * L^3;         % Units of mass

% ====== DIMENSIONLESS GROUPS ======

Fr = L / (T^2 * gCGS);
We = rhoCGS * L^3 / (T^2 * sigmaCGS);

% ====== NON-DIMENSIONALISE DATA ======

Lx = LxCGS / L;    % Dimensionless length of the domain
Tf = TfCGS / T;    % Dimensionless time of simulation

% ====== SPATIAL DISCRETISATION ======

dx = Lx / m;                      % Dimensionless mesh width
xVec = (-Lx/2 : dx : Lx/2)';

% ====== DtN OPERATOR ======

line = zeros(1, m + 1);
line(1) = 1 + 11/3;
line(2) = -16/9;
line(3) = -1/18;

for nn = 3:2:m
    an = -nn / (nn - 1) + (nn + 1/2) * log((nn + 1) / (nn - 1)) - 1;
    bn = -2 * nn * log((nn + 1) / (nn - 1)) + 4;
    cn = -nn / (nn + 1) + (nn - 1/2) * log((nn + 1) / (nn - 1)) - 1;

    line(nn) = line(nn) + an;
    line(nn + 1) = line(nn + 1) + bn;
    line(nn + 2) = line(nn + 2) + cn;
end

line = line / (pi * dx);

N = zeros(m + 1, m + 1);
for ii = 1:m + 1
    N(ii, :) = [fliplr(line(2:ii)), line(1:m - ii + 2)];
end

% ====== TEMPORAL DISCRETISATION ======

dt = Tf / Nt;
tVec = 0:dt:Tf;

% The appendix code constructs the centered second-difference matrix
% directly and uses it in the Crank-Nicholson block system.
DXX = diag(-2 * ones(1, m + 1)) + diag(ones(1, m), -1) + diag(ones(1, m), 1);

A = [[eye(m + 1), -dt * N / 2]; ...
     [dt * eye(m + 1) / (2 * Fr) - dt * DXX / (2 * We), eye(m + 1)]];

B = [[eye(m + 1), dt * N / 2]; ...
     [-dt * eye(m + 1) / (2 * Fr) + dt * DXX / (2 * We), eye(m + 1)]];

VecOld = zeros(2 * m + 2, 1);
Ps = zeros(m + 1, Nt + 1);

for kk = 1:length(tVec)
    Ps(:, kk) = exp(-xVec.^2) * (0.5 - 0.5 * cos(2 * pi * tVec(kk)));
end

outputDir = fullfile(fileparts(fileparts(mfilename('fullpath'))), 'results', 'figures');
if ~exist(outputDir, 'dir')
    mkdir(outputDir);
end

% ====== PRESSURE SURFACE ======

fig1 = figure('Visible', 'off');
surf(tVec, xVec, Ps, 'LineStyle', 'none', 'FaceColor', 'interp')
xlabel('t / \tau')
ylabel('x / l')
zlabel('p_s')
title('Gaussian pressure distribution')
exportgraphics(fig1, fullfile(outputDir, 'matlab_simulation_pressure_surface.png'))

% ====== SURFACE EVOLUTION ======

fig2 = figure('Visible', 'off');
plot(xVec, VecOld(1:m + 1), 'LineWidth', 2)
set(gca, 'ylim', [-0.02 0.02], 'xlim', [-Lx / 2, Lx / 2])
grid on
xlabel('x / l')
ylabel('\eta / l')
title(sprintf('t / \\tau = %.4f', tVec(1)))

snapshotSteps = [15, 55, 95, 135];
snapshotNames = { ...
    'matlab_simulation_snapshot_t15.png', ...
    'matlab_simulation_snapshot_t55.png', ...
    'matlab_simulation_snapshot_t95.png', ...
    'matlab_simulation_snapshot_t135.png'};

for kk = 1:Nt
    VecNew = A \ (B * VecOld + [zeros(m + 1, 1); -dt * (Ps(:, kk) + Ps(:, kk + 1))]);

    plot(xVec, VecNew(1:m + 1), 'LineWidth', 2)
    set(gca, 'ylim', [-0.02 0.02], 'xlim', [-Lx / 2, Lx / 2])
    grid on
    xlabel('x / l')
    ylabel('\eta / l')
    title(sprintf('t / \\tau = %.4f', tVec(kk + 1)))
    drawnow

    idx = find(snapshotSteps == kk, 1);
    if ~isempty(idx)
        exportgraphics(fig2, fullfile(outputDir, snapshotNames{idx}))
    end

    VecOld = VecNew;
end

File: src/simulation.py
"""
Crank-Nicholson simulation of a pressured free surface.

This module follows Chapter 5 and Appendix A of
"Impact of an Infinite Cylinder" and mirrors the MATLAB
Simulation.m listing included in the paper.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np

try:
    from src.dtn import build_dtn_matrix
except ModuleNotFoundError:
    import sys

    sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
    from src.dtn import build_dtn_matrix


@dataclass(frozen=True)
class SimulationParameters:
    lx_cgs: float = 160.0
    l_cgs: float = 1.0
    m: int = 2**10
    tau_cgs: float = 0.5
    tf_cgs: float = 0.5
    nt: int = 150
    rho_cgs: float = 1.0
    sigma_cgs: float = 70.0
    g_cgs: float = 980.0


def nondimensional_groups(params: SimulationParameters) -> tuple[float, float]:
    """Return the paper's Froude and Weber numbers."""
    length_unit = params.l_cgs
    time_unit = params.tau_cgs
    fr = length_unit / (time_unit**2 * params.g_cgs)
    we = params.rho_cgs * length_unit**3 / (time_unit**2 * params.sigma_cgs)
    return fr, we


def simulation_grid(params: SimulationParameters) -> tuple[float, float, np.ndarray, float, np.ndarray]:
    """Build the nondimensional spatial and temporal grids."""
    length_unit = params.l_cgs
    time_unit = params.tau_cgs

    lx = params.lx_cgs / length_unit
    tf = params.tf_cgs / time_unit

    dx = lx / params.m
    x_vec = np.arange(-lx / 2.0, lx / 2.0 + dx, dx)

    dt = tf / params.nt
    t_vec = np.arange(0.0, tf + dt, dt)

    return lx, tf, x_vec, dt, t_vec


def second_difference_matrix(m: int) -> np.ndarray:
    """
    Build the centered second-difference matrix used in Appendix A.

    This intentionally matches the paper's MATLAB listing.
    """
    dxx = np.diag(-2.0 * np.ones(m + 1))
    dxx += np.diag(np.ones(m), k=-1)
    dxx += np.diag(np.ones(m), k=1)
    return dxx


def gaussian_pressure(x_vec: np.ndarray, t_vec: np.ndarray) -> np.ndarray:
    """
    Pressure from equation (5.13) / Appendix A:

        p_s(x, t) = exp(-x^2) * (0.5 - 0.5 cos(2 pi t))
    """
    spatial = np.exp(-(x_vec**2))[:, None]
    temporal = (0.5 - 0.5 * np.cos(2.0 * np.pi * t_vec))[None, :]
    return spatial * temporal


def crank_nicholson_blocks(
    n_matrix: np.ndarray,
    dxx: np.ndarray,
    dt: float,
    fr: float,
    we: float,
) -> tuple[np.ndarray, np.ndarray]:
    """Build the block matrices A and B from equation (5.8)."""
    size = n_matrix.shape[0]
    eye = np.eye(size)

    a_mat = np.block(
        [
            [eye, -dt * n_matrix / 2.0],
            [dt * eye / (2.0 * fr) - dt * dxx / (2.0 * we), eye],
        ]
    )

    b_mat = np.block(
        [
            [eye, dt * n_matrix / 2.0],
            [-dt * eye / (2.0 * fr) + dt * dxx / (2.0 * we), eye],
        ]
    )

    return a_mat, b_mat


def run_simulation(
    params: SimulationParameters | None = None,
) -> dict[str, np.ndarray | float]:
    """Run the paper's free-surface simulation and return the computed fields."""
    params = params or SimulationParameters()

    fr, we = nondimensional_groups(params)
    lx, tf, x_vec, dt, t_vec = simulation_grid(params)
    n_matrix = build_dtn_matrix(params.m, lx / params.m)
    dxx = second_difference_matrix(params.m)
    a_mat, b_mat = crank_nicholson_blocks(n_matrix, dxx, dt, fr, we)
    ps = gaussian_pressure(x_vec, t_vec)

    vec_old = np.zeros(2 * (params.m + 1))
    eta_history = np.zeros((params.m + 1, params.nt + 1))
    phi_history = np.zeros((params.m + 1, params.nt + 1))

    eta_history[:, 0] = vec_old[: params.m + 1]
    phi_history[:, 0] = vec_old[params.m + 1 :]

    zero_block = np.zeros(params.m + 1)

    for kk in range(params.nt):
        forcing = np.concatenate([zero_block, -dt * (ps[:, kk] + ps[:, kk + 1])])
        rhs = b_mat @ vec_old + forcing
        vec_new = np.linalg.solve(a_mat, rhs)

        eta_history[:, kk + 1] = vec_new[: params.m + 1]
        phi_history[:, kk + 1] = vec_new[params.m + 1 :]
        vec_old = vec_new

    return {
        "Fr": fr,
        "We": we,
        "Lx": lx,
        "Tf": tf,
        "dt": dt,
        "x": x_vec,
        "t": t_vec,
        "pressure": ps,
        "eta": eta_history,
        "phi": phi_history,
    }


def save_outputs(results: dict[str, np.ndarray | float], output_dir: Path) -> None:
    """Save the pressure surface and the four paper snapshots."""
    output_dir.mkdir(parents=True, exist_ok=True)

    x_vec = np.asarray(results["x"])
    t_vec = np.asarray(results["t"])
    pressure = np.asarray(results["pressure"])
    eta = np.asarray(results["eta"])
    lx = float(results["Lx"])

    fig = plt.figure(figsize=(8, 5))
    ax = fig.add_subplot(111, projection="3d")
    tt, xx = np.meshgrid(t_vec, x_vec)

    # Match MATLAB's smooth surf rendering more closely by disabling the
    # default downsampling in Matplotlib's 3D surface plot.
    ax.plot_surface(
        tt,
        xx,
        pressure,
        cmap="viridis",
        linewidth=0,
        antialiased=False,
        rcount=pressure.shape[0],
        ccount=pressure.shape[1],
    )
    ax.set_xlabel("t / tau")
    ax.set_ylabel("x / l")
    ax.set_zlabel("p_s")
    ax.set_title("Gaussian pressure distribution")
    ax.view_init(elev=25, azim=-60)
    fig.tight_layout()
    fig.savefig(output_dir / "python_simulation_pressure_surface.png", dpi=180)
    plt.close(fig)

    snapshots = {
        15: "python_simulation_snapshot_t15.png",
        55: "python_simulation_snapshot_t55.png",
        95: "python_simulation_snapshot_t95.png",
        135: "python_simulation_snapshot_t135.png",
    }

    for step, filename in snapshots.items():
        fig, ax = plt.subplots(figsize=(8, 4))
        ax.plot(x_vec, eta[:, step], linewidth=2)
        ax.set_xlim([-lx / 2.0, lx / 2.0])
        ax.set_ylim([-0.02, 0.02])
        ax.grid(True)
        ax.set_xlabel("x / l")
        ax.set_ylabel("eta / l")
        ax.set_title(f"t / tau = {t_vec[step]:.4f}")
        fig.tight_layout()
        fig.savefig(output_dir / filename, dpi=180)
        plt.close(fig)


def main() -> None:
    results = run_simulation()
    repo_root = Path(__file__).resolve().parents[1]
    output_dir = repo_root / "results" / "figures"
    save_outputs(results, output_dir)


if __name__ == "__main__":
    main()
